### PR TITLE
[AIRFLOW-3605] Load plugins from entry_points

### PR DESCRIPTION
**Pull request by drewsonne**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. 
  - https://issues.apache.org/jira/browse/AIRFLOW-3605

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Added logic to load AirflowPlugins from the [entry_points](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#id16). Rather than moving files on the OS, this allows plugins to be installed and distributed via `pip`. 

Also added a callback to execute business logic from the plugin when the plugin is loaded.

### Tests

- [x] My PR adds the following unit tests:
 - `plugins_manager.PluginsEntrypointLoad`
 - `plugins_manager.PluginsTestValidator`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

~I haven't written the documentation yet, as I wanted to see if there was an appetite for this kind of loading of plugins.~

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by mik-laj** [comment]
Already someone has previously dealt with this topic:
https://github.com/apache/incubator-airflow/pull/730
https://issues.apache.org/jira/browse/AIRFLOW-594

**Comment by drewsonne** [comment]
@mik-laj Ah, I missed that in my search. Looks like my logic is remarkably similar, so which is good to know 😄. I noticed that the other PR includes DAGs as well; that's a great idea. I think for simplicity, I'd keep this smaller PR, and go back to the DAGs as packages at a later date, as I also would find this very useful.

I shall steal the documentation (and attribute it) from that PR though.

**Comment by feng-tao** [comment]
@drewsonne , this is an interesting change. I will take a look.

File: airflow/plugins_manager.py [hunk 1]
```diff
@@ -90,19 +142,17 @@ def validate(cls):
...
+                if is_valid_plugin(obj, plugins):
+                    plugins.append(obj)
 
         except Exception as e:
             log.exception(e)
             log.error('Failed to import plugin %s', filepath)
             import_errors[filepath] = str(e)
 
+plugins += load_entrypoint_plugins(
+    pkg_resources.iter_entry_points('airflow.plugins'))
```
Comment (ashb) [inline airflow/plugins_manager.py]: Should the currently loaded plugins be passed in to `load_entrypoint_plugs()`? I.e something like:

```python
plugins += load_entrypoint_plugins(
    pkg_resources.iter_entry_points('airflow.plugins'),
    plugins)
```

File: docs/plugins.rst [hunk 2]
```diff
@@ -220,3 +220,26 @@ Note on role based views
...
+    setup(
+        name="my-package",
+        ...
+        entry_points = {
+            'airflow.plugins': [
+                'my_plugin = my_package.my_plugin:MyAirflowPlugin'
+            ]
+        }
+    )
+
```
Comment (ashb) [inline docs/plugins.rst:270]: Could you extend this doc to make it clear under what name this plugin will be available? Because I don't think that _any_ of these names actually appears in the airflow plugin import statement at use time? (Though that might be clearer if it did?)

Comment (drewsonne) [reply docs/plugins.rst:270]: I'll expand the documentation when I get home, but given a class like

```python
# my_package/my_plugin.py
from airflow.plugins_manager import AirflowPlugin

class MyAirflowPlugin(AirflowPlugin):
    name = 'my_namespace.MyPlugin'
```
I think (I can't remember off the top of my head, but I will test) that the plugin would be exposed under `airflow.operators.my_namespace.MyPlugin`, `airflow.sensors.my_namespace.MyPluigin`, etc.

Comment (drewsonne) [reply docs/plugins.rst:270]: I'd write some test cases for this functionality, but it is a bit difficult given the structure of this file. Would there be objection to wrapping this functionality in some functions as I did with `is_valid_plugin(...)` to make this a bit easier to test?

File: airflow/plugins_manager.py (hunk 1 above)
Comment (drewsonne) [reply airflow/plugins_manager.py]: You are correct, we're not checking for double loading of plugins from both sources. Will change.

File: docs/plugins.rst (hunk 2 above)
Comment (drewsonne) [reply docs/plugins.rst:270]: Before I spent time working on testcases, this would be the idea (and I'd put it in a separate PR) https://github.com/drewsonne/incubator-airflow/pull/1/files

Comment (ashb) [reply docs/plugins.rst:270]: The class as you have it in that PR1 you link to looks fine from a very quick glance.

I was more thinking that in the setup.py example in the docs that the `my_plugin` (and maybe `MyAirflowPlugin`?) name doesn't matter - it's not part of the the "name" that is exposed by airflow is it? It would be `from airflow.operators.my_namespace import MyOperator` right?

Comment (drewsonne) [reply docs/plugins.rst:270]: Ah right, i have not used the “name” of the entry point as it is already handled by the ‘AirflowPlugin.name’. We could use it, but it seems like a precedence would just complicate things.

Comment (drewsonne) [reply docs/plugins.rst:270]: And yes, that import is correct. It's using the same namespace building logic as the file based plugins, so it shouldn't change beyond that.

**Comment by feng-tao** [comment]
@ashb, FYI, plugin_mangager test never get executed in CI. I am trying to fixing it in https://github.com/apache/airflow/pull/4464. Please wait for that pr to check in before merging this one.

**Comment by drewsonne** [comment]
@feng-tao I've got a PR to your airflow repo, which fixes one of the failures in the test cases https://github.com/feng-tao/incubator-airflow/pull/12 for #4464

**Comment by feng-tao** [comment]
@drewsonne @ashb , the plugin-manager test has been fixed. Please rebase your pr with master. Thanks.

**Review by ashb — APPROVED** [review]
Code looks good, I haven't looked at why tests are failing thought :)

**Comment by drewsonne** [comment]
Uhhhhh... it's failing on writing a log 😕 

https://travis-ci.org/apache/airflow/jobs/477803046#L4897
```======================================================================
37) ERROR: test_failed_load_entrpoint_plugins (tests.plugins.test_plugins_manager_www.PluginsTestEntrypointLoad)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/plugins/test_plugins_manager_www.py line 128 in test_failed_load_entrpoint_plugins
      )], []),
    airflow/plugins_manager.py line 88 in load_entrypoint_plugins
      log.debug('Importing entry_point plugin %s', entry_point.name)
    /usr/lib/python2.7/logging/__init__.py line 1155 in debug
      self._log(DEBUG, msg, args, **kwargs)
    /usr/lib/python2.7/logging/__init__.py line 1286 in _log
      self.handle(record)
    /usr/lib/python2.7/logging/__init__.py line 1296 in handle
      self.callHandlers(record)
    /usr/lib/python2.7/logging/__init__.py line 1336 in callHandlers
      hdlr.handle(record)
    /usr/lib/python2.7/logging/__init__.py line 759 in handle
      self.emit(record)
    .tox/py27-backend_mysql-env_docker/lib/python2.7/site-packages/nose/plugins/logcapture.py line 82 in emit
      self.buffer.append(self.format(record))
    /usr/lib/python2.7/logging/__init__.py line 734 in format
      return fmt.format(record)
    /usr/lib/python2.7/logging/__init__.py line 465 in format
      record.message = record.getMessage()
    /usr/lib/python2.7/logging/__init__.py line 329 in getMessage
      msg = msg % self.args
   TypeError: coercing to Unicode: need string or buffer, MagicMock found```

Initial inspection looks like I'm using logging the same as everywhere else, but I will have to have a closer look later. It looks like a problem with the method of test implementation rather than the code itself... I think.

**Comment by feng-tao** [comment]
I restart the fail jenkins CI.

**Comment by drewsonne** [comment]
Ok, this looks good from my side now. The codecov says it's reducing coverage, but in files this PR doesn't touch, so I'm not sure why :-/

Anything else or is this good to go?

File: tests/plugins/test_plugins_manager_www.py [hunk 3]
```diff
@@ -83,3 +85,66 @@ def test_menu_links(self):
         [menu_link] = [ml for ml in category.get_children()
                        if isinstance(ml, MenuLink)]
         self.assertEqual('Test Menu Link', menu_link.name)
+
+
+class MockPluginA(AirflowPlugin):
```
Comment (feng-tao) [inline tests/plugins/test_plugins_manager_www.py]: sorry, but could you move the MockPlugin to test_plugin file which is a centralized place for all the test plugins?

File: tests/plugins/test_plugins_manager_www.py [hunk 4]
```diff
@@ -83,3 +85,66 @@ def test_menu_links(self):
...
+
+class MockPluginB(AirflowPlugin):
+    name = 'plugin-b'
+
+
+class MockPluginC(AirflowPlugin):
+    name = 'plugin-c'
+
+
+class PluginsTestEntrypointLoad(unittest.TestCase):
```
Comment (feng-tao) [inline tests/plugins/test_plugins_manager_www.py:91]: are these two unit test only worked for old UI? If not I would suggest to move to test_plugins_manager_rbac.py once we delete the old UI code, we will remove this unit test file as well.

**Review by feng-tao — APPROVED** [review]
Two small nits, otherwise lgtm

File: tests/plugins/test_plugins_manager_www.py (hunk 4 above)
Comment (drewsonne) [reply tests/plugins/test_plugins_manager_www.py:91]: I'm not sure, as I haven't touched on the UI parts of the code base yet.

This code is not really anything to do with the UI as far as I'm aware. Or at least, it's not specific to one UI or the other. When I pulled the above PR changes in, this is where my code was move to, so the old test/plugin_manager.py file was renamed to this.

tl;dr it's not UI specific(?), so I'm not sure where it goes.

File: tests/plugins/test_plugins_manager_www.py (hunk 3 above)
Comment (drewsonne) [reply tests/plugins/test_plugins_manager_www.py]: Done, I just have to wait for the tests to run again

**Comment by feng-tao** [comment]
@drewsonne , FYI, two flake8 errors:
```flake8 runtests: commands[0] | flake8
./tests/plugins/test_plugins_manager_www.py:33:1: F401 'airflow.plugins_manager.AirflowPlugin' imported but unused
./tests/plugins/test_plugins_manager_www.py:90:1: E302 expected 2 blank lines, found 1
```

**Comment by drewsonne** [comment]
Argh, stupid mistake. Here's a PR to try to shorten the turnaround time for travis alerting on lint failures
https://github.com/apache/airflow/pull/4477/files

**Comment by feng-tao** [comment]
thanks @drewsonne , merging now.

File: tests/plugins/test_plugins_manager_www.py (hunk 4 above)
Comment (feng-tao) [reply tests/plugins/test_plugins_manager_www.py:91]: I guess we could leave it as it is now. But FYI, the old UI and new UI depends on different flask framework which works a bit differently with the plugin manager(e.g. old UI could integrate with blueprint while old UI could not etc). But tldr: let's leave it as it is now.

**Comment by drewsonne** [comment]
@feng-tao amazing, thank you! looking forward to the release for this 😄

**Comment by feng-tao** [comment]
cc @kaxil  as he is driving the 10.2 release.

**Comment by kaxil** [comment]
Awesome work @drewsonne .  Included it in 1.10.2

File: tests/plugins/test_plugins_manager_www.py (hunk 4 above)
Comment (drewsonne) [reply tests/plugins/test_plugins_manager_www.py:91]: From what I've seen so far, the plugin manager loads the plugins, and dumps their respective components (hooks, operators, blueprints, etc) into lists and then it's up to other components to decide which lists to reference.

I'm happy to put in another PR to move it so keep thing Clean™, but I'm not sure where it'd go. Are there plans for the old UI to be deprecated at a time in the future? That could solve the problem.
